text/0000-ide.md

```diff
@@ -0,0 +1,530 @@
+further errors (I presume that IDEs provide a mechanism for the compiler to
+communicate these asynchronously to the IDE plugin).
+
+In addition we must produce data to update the oracle, this should be done
+directly, without involving the IDE plugin.
+
+Quick-check does not generate executable code or crate metadata. However, it
+should (probably) update the metadata used for incremental compilation.
+
+
+### Multiple crates
+
+Quick check only applies to a single crate, however, after some changes we might
+need to re-compile dependent crates. This is the IDE's responsibility. In the
+short term we can just trigger a full re-build (via Cargo) when the user starts
+editing a file belonging to a different crate (there will obviously be some lag
+there). The compiler must also generate crate metadata for the modified crate.
+
+Long term, the IDE might keep track of the dependency graph between crates
+(provided by Cargo). The quick-check should signal when a crate's public
+interface changes due to re-compilation. In that case the IDE can trigger
+background re-compilation of dependent crates (possibly with some
+delay/batching).
+
+
+## The Oracle
+
+The oracle is a long-running tool which takes input from both full builds and
+quick-checks, and responds to queries about a Rust program. Of particular note
+is that it knows about a whole project, not just a single crate. In fact, other
+than as a kind of module, it doesn't much care about the notion of a crate at
+all.
+
+We require a data format for getting metadata from the compiler to the oracle.
+Unfortunately none of the existing ones are quite right. Crate metadata is not
+complete enough (it mostly only contains data about interfaces, not function
+bodies), save-analysis data has been processed too far (basically into strings)
+which loses some of the structure that would be useful, debuginfo is not Rust-
+centric enough (i.e., does not contain Rust type information) and is based on
+expanded source code. Furthermore, serialising any of the compiler's IRs is not
+good enough: the AST and HIR do not contain any type or name information, the
+HIR and MIR are post-expansion.
+
+The best option seems to be the save-analysis information. This is in a poor
+format, but is the 'right' data (it can be based on an early AST and includes
+type and name information). It can be evolved to be more efficient form over the
+long run (it has been a low priority task for a while to support different
+formats for the information).
+
+Full builds will generate a dump of save-analysis data for a whole crate. Quick
+checks will generate data for the changed code. In both cases the oracle must
+incrementally update its knowledge of the source code. How exactly to do this
+when neither names nor ids are stable is an interesting question, but too much
+detail for this RFC (especially as the implementation of ids in the compiler is
+evolving).
```

Comment:
It could, I list this as an alternative in the alternatives section (I assume you mean the quick-check version of rustc). I think it may be a better approach. The only real downside is that the oracle then has to know about dependencies between crates.

In terms of when to do a full build of a crate, I think you want to avoid this as much as possible, so the IDE is the best driver to choose when it is necessary.